### PR TITLE
[SPARK-34297][SQL][SS] Add metrics for data loss and offset out range for KafkaMicroBatchStream

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch proposes to add a couple of metrics in scan node for Kafka batch streaming query.

### Why are the changes needed?

When testing SS, I found it is hard to track data loss of SS reading from Kafka. The micro batch scan node has only one metric, number of output rows. Users have no idea how many offsets to fetch are out of Kafka, how many times data loss happens. These metrics are important for users to know the quality of SS query running.

### Does this PR introduce _any_ user-facing change?

Yes, adding two metrics to micro batch scan node for Kafka batch streaming.

### How was this patch tested?

Currently I tested on internal cluster with Kafka:

Screen Shot 2021-04-22 at 7 16 29 PM

I was trying to add unit test. But as our batch streaming query disallows to specify ending offsets. If I only specify an out-of-range starting offset, when we get offset range in `getRanges`,  any negative size range will be filtered out. So it cannot actually test the case of fetched non-existing offset.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @HeartSaVioR, @gaborgsomogyi, @xuanyuanking FYI

**Comment by HeartSaVioR** [comment]
I think the most important part of this PR is the addition of metrics in public API. As the addition is going to affect the public API, I think we'll need to ping folks working on DSv2. Probably even easier to make such API changes to be separated one, as this is no longer simply related to SS.

**Comment by viirya** [comment]
Yes, I agree. This is not only for related to SS, there is SQL change especially in the public API.

**Comment by viirya** [comment]
cc @cloud-fan @aokolnychyi @dongjoon-hyun @rdblue @sunchao too

**Comment by cloud-fan** [comment]
The feature makes a lot of sense, but to me building the infra to report data source metrics is the most important thing here. @viirya can you add more details about the DS v2 API design, how it works with batch and streaming, how it is integrated into the batch and streaming execution?

**Comment by viirya** [comment]
One public interface `CustomMetric` is added. The interface simply can return metric name, description and value.

Currently there is only metric implementation `CustomSumMetric`. Normally we will have other metrics like `CustomSizeMetric` and `CustomTimingMetric`, corresponding to SQL size and timing metrics. But for the SS metrics I added, I only use sum metric, so I don't add them.

There are two public methods added to existing public interfaces.

* PartitionReader.getCustomMetrics(): returns an array of `CustomMetric`. Here is where the actual metrics values are collected.
* MicroBatchStream.supportedCustomMetrics(): returns an array of supported custom metrics with name and description. Basically `MicroBatchStream` is for micro-batch and `ContinuousStream` is for continuous streaming. I only test micro-batch for now.

The metric collection happens as following.

1. `MicroBatchScanExec` calls its stream's `supportedCustomMetrics` method to know what custom SQL metrics should be added.
2. `MicroBatchScanExec` passes a callback function for updating metrics to `DataSourceRDD`. When `DataSourceRDD` completes data consumption, it calls the callback function.

Note that currently some changes are specific for SS micro-batch. If there is a consensus we should extend to general DS v2, I need to make corresponding change. For example, moving `supportedCustomMetrics` from `MicroBatchStream` to general `Scan`. So both batch, streaming (micro-bath, continuous streaming) can use it.

So for general metrics report, I think it works like:

1. `BatchScanExec`, `MicroBatchScanExec`, `ContinuousScanExec` call its scan's `supportedCustomMetrics` method to know what custom SQL metrics should be added.
2. `BatchScanExec`, `MicroBatchScanExec` pass a callback function for updating metrics to `DataSourceRDD`. When `DataSourceRDD` completes data consumption, it calls the callback function. For `ContinuousScanExec`, it passes similar callback function to `ContinuousDataSourceRDD` for updating metrics.

**Comment by viirya** [comment]
To facilitate the discuss and review, I may create another PR including only DS v2 API change without SS-specific change.

**Comment by cloud-fan** [comment]
Yea creating a separate PR SGTM. Let's have a high-level discussion first (I haven't read this PR yet). From my understanding, metrics in batch execution can be done as:
1. data source returns custom metrics in each read/write task (via task completion event or heartbeat event)
2. data source aggregates custom metrics from all tasks

For microbatch execution, we just repeat the batch execution steps for each microbatch. And we update the metrics in the UI in every microbatch.

For continuous execution, it's like an endless batch execution, so we can only use heartbeat events to update metrics. And we update the metrics in the UI in every epoch.

The problem here is how to integrate this with Spark SQL. One idea is to use `AccumulatorV2`, which is a public API already and is very flexible. But we need to figure out how to make it work with the SQL UI. The other idea is to use `SQLMetrics`, which is private and we need some API design to map public API to `SQLMetrics`. It also limits the way of aggregating metrics.

**Comment by viirya** [comment]
I extract generalized DS v2 change to #31451.

**Comment by viirya** [comment]
> For continuous execution, it's like an endless batch execution, so we can only use heartbeat events to update metrics. And we update the metrics in the UI in every epoch.

Because I only test micro-batch case, I'm not sure if SQLMetrics metrics work for continuous streaming or not. But I also don't see special handling of common metrics (number of rows) in continuous streaming. I will check if SQLMetrics work with continuous streaming.

**Comment by viirya** [comment]
> The problem here is how to integrate this with Spark SQL. One idea is to use `AccumulatorV2`, which is a public API already and is very flexible. But we need to figure out how to make it work with the SQL UI. The other idea is to use `SQLMetrics`, which is private and we need some API design to map public API to `SQLMetrics`. It also limits the way of aggregating metrics.

Currently I took the second idea. I add one public interface `CustomMetric` to map to `SQLMetrics`. There will be corresponding metric classes such as `CustomSumMetric`. This approach is easier to work with SQL UI.

**Comment by viirya** [comment]
BTW, I will re-test this on internal cluster later.

**Comment by viirya** [comment]
I haver verified the updated change on internal Spark cluster. Captured screenshot is in the description.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/CustomMetrics.scala [hunk 1]
```diff
@@ -47,7 +47,8 @@ object CustomMetrics {
 }
 
 /**
- * Built-in `CustomMetric` that sums up metric values.
+ * Built-in `CustomMetric` that sums up metric values. Note that please extend this class
+ * and override `name` and `description` to create your custom metric for real usage.
  */
 class CustomSumMetric extends CustomMetric {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/CustomMetrics.scala:53]: shall we make it an abstract class and force people to implement name and description?

**Review by cloud-fan — APPROVED** [review]
LGTM except one comment

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/CustomMetrics.scala (hunk 1 above)
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/metric/CustomMetrics.scala:53]: okay. had the same thought.

**Comment by viirya** [comment]
Thanks. Merging to master.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/metric/CustomMetrics.scala [hunk 2]
```diff
@@ -47,25 +47,21 @@ object CustomMetrics {
...
 /**
- * Built-in `CustomMetric` that sums up metric values.
+ * Built-in `CustomMetric` that sums up metric values. Note that please extend this class
+ * and override `name` and `description` to create your custom metric for real usage.
  */
-class CustomSumMetric extends CustomMetric {
-  override def name(): String = "CustomSumMetric"
-
-  override def description(): String = "Sum up CustomMetric"
+abstract class CustomSumMetric extends CustomMetric {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/metric/CustomMetrics.scala:53]: This is in the internal package. Shall we move it to the v2 package and rewrite it using java?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/metric/CustomMetrics.scala:53]: or we plan to use it for internal sources only?

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/metric/CustomMetrics.scala:53]: I will create a followup to do it.

**Pull request closed without merging**